packages/react-router-dom/index.tsx

```diff
@@ -477,17 +493,62 @@ export function RouterProvider({
         }
       });
 
-      if (
-        !viewTransitionOpts ||
+      let isViewTransitionUnavailable =
         router.window == null ||
-        typeof router.window.document.startViewTransition !== "function"
-      ) {
-        // Mid-navigation state update, or startViewTransition isn't available
-        optInStartTransition(() => setStateImpl(newState));
-      } else if (transition && renderDfd) {
+        typeof router.window.document.startViewTransition !== "function";
+
+      // If this isn't a view transition or it's not available in this browser,
+      // just update and be done with it
+      if (!viewTransitionOpts || isViewTransitionUnavailable) {
+        if (flushSync) {
+          flushSyncSafe(() => setStateImpl(newState));
+        } else {
+          optInStartTransition(() => setStateImpl(newState));
+        }
+        return;
+      }
+
+      // flushSync + startViewTransition
+      if (flushSync) {
```

Comment:
I don't know if I could combine these two flows or not code-wise but for the first pass I just kept `startViewTransition`/`flushSync` entirely separate from `startViewTransition`/`startTransition`.  They use the same local states, but the series of sequential effects are solely for `startTransition`, while `flushSync` does it all here - bang bang bang